File: fsharp/phone-number/PhoneNumber.fs
module PhoneNumber

let isInvalidPunctuation c =
  let validPunctuation = ['+'; '('; ')'; '-'; '.';]
  System.Char.IsPunctuation c && not <| List.contains c validPunctuation

let hasAlphanumerics (nums: array<char>) =
  Array.exists System.Char.IsLetter nums

let hasInvalidPunctuation (nums: array<char>) =
  Array.exists isInvalidPunctuation nums

let isLessThanTenDigits (nums: array<char>) =
  Array.length nums < 10

let isMoreThanElevenDigits (nums: array<char>) =
  Array.length nums > 11

let countryCodeIsInvalid (nums: array<char>) =
  Array.length nums = 11 && Array.get nums 0 <> '1'

let areaCodeStartsWithZero (nums: array<char>) =
  Array.get nums (Array.length nums - 10) = '0'

let areaCodeStartsWithOne (nums: array<char>) =
  Array.get nums (Array.length nums - 10) = '1'

let exchangeCodeStartsWithZero (nums: array<char>) =
  Array.get nums (Array.length nums - 7) = '0'

let exchangeCodeStartsWithOne (nums: array<char>) =
  Array.get nums (Array.length nums - 7) = '1'

let clean (input: string) =
  let numsArray = Array.ofSeq input
  match numsArray with
  | numsArray when hasAlphanumerics numsArray -> Error "alphanumerics not permitted"
  | numsArray when hasInvalidPunctuation numsArray -> Error "punctuations not permitted"
  | _ ->
    let cleanedNums = Array.filter System.Char.IsNumber numsArray
    match cleanedNums with
    | cleanedNums when isLessThanTenDigits cleanedNums -> Error "incorrect number of digits"
    | cleanedNums when isMoreThanElevenDigits cleanedNums -> Error "more than 11 digits"
    | cleanedNums when countryCodeIsInvalid cleanedNums -> Error "11 digits must start with 1"
    | cleanedNums when areaCodeStartsWithZero cleanedNums -> Error "area code cannot start with zero"
    | cleanedNums when areaCodeStartsWithOne cleanedNums -> Error "area code cannot start with one"
    | cleanedNums when exchangeCodeStartsWithZero cleanedNums -> Error "exchange code cannot start with zero"
    | cleanedNums when exchangeCodeStartsWithOne cleanedNums -> Error "exchange code cannot start with one"
    | _ ->
      Ok (System.UInt64.Parse (System.String (Array.sub cleanedNums ((Array.length cleanedNums) - 10) 10)))

File: javascript/phone-number/phone-number.js
//
// This is only a SKELETON file for the 'Phone Number' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

const VALID_NUMBER_LENGTH = 10;
const MAX_NUMBER_LENGTH = 11;
const COUNTRY_CODE_INDEX = 0;
const AREA_CODE_INDEX = 0;
const EXCHANGE_CODE_INDEX = 3;

export const clean = phoneNumber => {
  const cleanedNumber = phoneNumber.replace(/[\s\.\-\(\)\+]/g, '');
  if (/[A-Za-z]/.test(cleanedNumber)) throw new Error('Letters not permitted');
  if (/\D/.test(cleanedNumber)) throw new Error('Punctuations not permitted');
  if (cleanedNumber.length < VALID_NUMBER_LENGTH) throw new Error('Incorrect number of digits');
  if (cleanedNumber.length > MAX_NUMBER_LENGTH) throw new Error(`More than ${MAX_NUMBER_LENGTH} digits`);
  if (cleanedNumber.length === MAX_NUMBER_LENGTH && cleanedNumber[COUNTRY_CODE_INDEX] !== '1')
    throw new Error(`${MAX_NUMBER_LENGTH} digits must start with 1`);

  const cleanedNumberLessCountry = cleanedNumber.slice(-VALID_NUMBER_LENGTH);
  if (cleanedNumberLessCountry[AREA_CODE_INDEX] === '0') throw new Error('Area code cannot start with zero');
  if (cleanedNumberLessCountry[AREA_CODE_INDEX] === '1') throw new Error('Area code cannot start with one');
  if (cleanedNumberLessCountry[EXCHANGE_CODE_INDEX] === '0') throw new Error('Exchange code cannot start with zero');
  if (cleanedNumberLessCountry[EXCHANGE_CODE_INDEX] === '1') throw new Error('Exchange code cannot start with one');

  return cleanedNumberLessCountry;
}